File: backend/dotnet/src/Mateu.Core/YamlSpecLoader.cs
using System.Collections.Concurrent;
using Mateu.Uidl;

namespace Mateu.Core;

/// <summary>
/// Loads a page defined in a YAML file under <c>specs/ui/</c> relative to the working directory —
/// the .NET analogue of Java's classpath-based YamlUidlLoader. A spec is a component tree, optionally
/// wrapped in an envelope with a <c>modelView:</c> key naming the logic class that supplies state and
/// actions (the YAML supplies only the layout). The binding is by convention, as everywhere in Mateu:
/// a FormField <c>id="name"</c> binds to the ModelView's <c>Name</c> property, a Button
/// <c>actionId="save"</c> to its <c>Save()</c> method.
/// </summary>
/// <remarks>
/// Specs are static files, so each route is parsed once and cached (a miss is cached too, so an
/// unmatched route — checked on every request that has no view class — does not stat the disk each
/// time). Editing a spec during development needs a restart to be picked up. Override the specs
/// directory with the MATEU_SPECS_DIR environment variable (default: <c>specs/ui</c> under the cwd).
/// </remarks>
public sealed class YamlSpecLoader
{
    /// <summary>A parsed page spec: the layout, plus the ModelView class name when the YAML declares one.</summary>
    public sealed record Spec(string? ModelView, IComponent? Layout);

    private static readonly Spec None = new(null, null);
    private readonly ConcurrentDictionary<string, Spec> _byRoute = new();
    private readonly string _dir;

    /// <summary>When a route's registry entry names a <c>definition</c>, THAT file is the layout —
    /// instead of the <c>&lt;route&gt;.yaml</c> convention, which ties a screen's layout to its URL
    /// and so prevents one definition from serving several routes.</summary>
    private readonly RouteRegistry _registry;

    public YamlSpecLoader(string? dir = null, RouteRegistry? registry = null)
    {
        _dir = dir ?? Environment.GetEnvironmentVariable("MATEU_SPECS_DIR")
                   ?? Path.Combine("specs", "ui");
        _registry = registry ?? new RouteRegistry(_dir);
    }

    /// <summary>The parsed spec for a route (<c>specs/ui/&lt;route&gt;.yaml</c>), or null when there is none.</summary>
    public Spec? LoadSpec(string? route)
    {
        var spec = _byRoute.GetOrAdd(Normalize(route), Parse);
        return ReferenceEquals(spec, None) ? null : spec;
    }

    private Spec Parse(string normalizedRoute)
    {
        var entry = _registry.Match(normalizedRoute)?.Entry;
        var declared = string.IsNullOrWhiteSpace(entry?.Definition) ? null : entry!.Definition;
        var path = Path.Combine(_dir, declared ?? normalizedRoute + ".yaml");
        if (!File.Exists(path)) return None;
        try
        {
            var (modelView, layout) = YamlComponentBuilder.ParseSpec(File.ReadAllText(path));
            if (layout is null) return None;
            // The definition is layout; the binding to a view model belongs to the route entry. A
            // YAML that still declares modelView: keeps working and wins — but a definition shared
            // by several routes must NOT name one, or it could only ever serve the class it names.
            if (string.IsNullOrWhiteSpace(modelView) && !string.IsNullOrWhiteSpace(entry?.ViewModel))
                modelView = entry!.ViewModel;
            return new Spec(modelView, layout);
        }
        catch { return None; }
    }

    private static string Normalize(string? route)
    {
        var r = route ?? "";
        var q = r.IndexOf('?');
        if (q >= 0) r = r[..q];
        r = r.Trim('/');
        return r is "_empty" or "_no_route" or "_no_home_route" ? "" : r;
    }
}

File: backend/python/mateu_core/yaml_spec_loader.py
"""Loads a page defined in a YAML file under ``specs/ui/`` relative to the working directory — the
Python analogue of Java's classpath-based YamlUidlLoader and .NET's YamlSpecLoader.

A spec is a component tree, optionally wrapped in an envelope with a ``modelView:`` key naming the
logic class that supplies state and actions (the YAML supplies only the layout). The binding is by
convention, as everywhere in Mateu: a FormField ``id="name"`` binds to the ModelView's ``name``
attribute, a Button ``actionId="save"`` to its ``save()`` method.

Specs are static files, so each route is parsed once and cached (a miss too, so an unmatched route —
checked on every request that has no view class — does not stat the disk each time). Editing a spec
during development needs a restart to be picked up. Override the directory with the
``MATEU_SPECS_DIR`` environment variable (default: ``specs/ui`` under the cwd).
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

from mateu_core.yaml_preview import parse_spec


@dataclass(frozen=True)
class Spec:
    """A parsed page spec: the layout, plus the ModelView class name when the YAML declares one."""

    model_view: str | None
    layout: object | None


class YamlSpecLoader:
    def __init__(self, directory: str | None = None) -> None:
        self._dir = Path(directory or os.environ.get("MATEU_SPECS_DIR") or Path("specs") / "ui")
        self._by_route: dict[str, Spec | None] = {}

    def load_spec(self, route: str | None) -> Spec | None:
        key = _normalize(route)
        if key not in self._by_route:
            self._by_route[key] = self._parse(key)
        return self._by_route[key]

    def _parse(self, normalized_route: str) -> Spec | None:
        path = self._dir / f"{normalized_route}.yaml"
        if not path.is_file():
            return None
        try:
            model_view, layout = parse_spec(path.read_text())
        except OSError:
            return None
        return Spec(model_view, layout) if layout is not None else None


def _normalize(route: str | None) -> str:
    r = route or ""
    q = r.find("?")
    if q >= 0:
        r = r[:q]
    r = r.strip("/")
    return "" if r in ("_empty", "_no_route", "_no_home_route") else r
